tool/gravity/cli/logging.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"fmt"
+	"log/syslog"
+	"strings"
+
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/utils"
+	"github.com/gravitational/gravity/lib/utils/cli"
+
+	"github.com/gravitational/trace"
+)
+
+// Executable executes a gravity command.
+type Executable func() error
+
+// CmdExecer handles execution of a gravity command.
+type CmdExecer struct {
+	// Exe specifies an executable gravity command.
+	Exe Executable
+	// Parser specifies the gravity arguments parser function.
+	Parser cli.ArgsParserFunc
+	// Args specifies the provided gravity command arguments.
+	Args []string
+	// ExtraArgs specifies the provided extra arguments.
+	ExtraArgs []string
+}
+
+// Execute executes the gravity command while logging the start and completion
+// of the command.
+func (r *CmdExecer) Execute() (err error) {
+	sanitizedCmd, err := r.redactCmd()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	cmdString := fmt.Sprintf("%s -- %s", strings.Join(sanitizedCmd, " "), strings.Join(r.ExtraArgs, " "))
+
+	logEntry(fmt.Sprintf("[RUNNING]: %s", cmdString))
+	defer func() {
+		if r := recover(); r != nil {
+			logEntry(fmt.Sprintf("[FAILURE]: %s: [PANIC]: %v", cmdString, r))
```

Comment:
I don't think we should be silencing a panic. It might be worthwhile to capture to report the failing command but I'd still want to see the stack so either re-panic or log `string(debug.Stack())` to get stack (since recover runs in the context of the failing function).

Reply:
Ah, did not intend to silence the panic. Updated to re-panic after logging.

Reply:
nit: `[PANIC]: ...` will repeat what the std panic handling also dumps to the console.

Reply:
I don't quite understand the comment. Could you clarify?

Reply:
Say you have an index out of bounds somewhere. The code as written will output the following:
```sh
[FAILURE]: [/path/tobinary args...] [PANIC]: runtime error: index out of range [0] with length 0.
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
...
```
The same panic message is output 3 times in this case.

Reply:
Oh, I see. I think this okay. The first line `[FAILURE] ...] will only be written to the system journal. The panic message will still be output twice, but this seems like the standard way of re-panicking.